lib/core/src/Cardano/Wallet/DB/SqliteTypes.hs

```diff
@@ -0,0 +1,257 @@
+{-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+
+module Cardano.Wallet.DB.SqliteTypes where
+
+import Prelude
+
+import Control.Monad
+    ( (>=>) )
+import Data.Aeson
+import Data.Aeson.Types
+    ( Parser )
+import Data.Bifunctor
+    ( bimap, first )
+import Data.Proxy
+    ( Proxy (..) )
+import Data.Quantity
+    ( Quantity (..), getPercentage, mkPercentage )
+import Data.Text
+    ( Text )
+import Data.Text.Class
+    ( FromText (..)
+    , TextDecodingError (..)
+    , ToText (..)
+    , fromTextMaybe
+    , getTextDecodingError
+    )
+import Data.Word
+    ( Word64, Word8 )
+import Database.Persist.Sqlite
+    ( PersistField (..), PersistFieldSql (..) )
+import GHC.Generics
+    ( Generic )
+import Web.HttpApiData
+import Web.PathPieces
+
+import qualified Data.Text as T
+
+import Cardano.Wallet.Primitive.Types
+    ( Coin (..)
+    , Direction (..)
+    , Hash (..)
+    , SlotId (..)
+    , TxStatus (..)
+    , WalletId (..)
+    , WalletState (..)
+    , flatSlot
+    , fromFlatSlot
+    , isValidCoin
+    )
+
+----------------------------------------------------------------------------
+-- Helper functions
+
+parseUrlPieceFromText :: FromText a => Text -> Either Text a
+parseUrlPieceFromText = first (T.pack . getTextDecodingError) . fromText
+
+aesonFromText :: FromText a => String -> Value -> Parser a
+aesonFromText what = withText what $ either (fail . show) pure . fromText
+
+----------------------------------------------------------------------------
+-- Direction
+
+instance PersistField Direction where
+    toPersistValue = toPersistValue . directionToBool
+    fromPersistValue pv = do
+        let err = "not a valid value: " <> T.pack (show pv)
+        bimap (const err) directionFromBool (fromPersistValue pv)
+
+instance PersistFieldSql Direction where
+    sqlType _ = sqlType (Proxy @Bool)
+
+directionToBool :: Direction -> Bool
+directionToBool Incoming = True
+directionToBool Outgoing = False
+
+directionFromBool :: Bool -> Direction
+directionFromBool True = Incoming
+directionFromBool False = Outgoing
+
+----------------------------------------------------------------------------
+-- WalletId
+
+instance PersistField WalletId where
+    toPersistValue =
+        toPersistValue . toText
+    fromPersistValue pv = do
+        a <- fromText <$> fromPersistValue pv
+        case a of
+             Left _     ->
+                 Left $ "not a valid value: " <> T.pack (show pv)
+             Right txid ->
+                 pure txid
+
+instance PersistFieldSql WalletId where
+    sqlType _ = sqlType (Proxy @Text)
+
+instance Read WalletId where
+    readsPrec _ = error "readsPrec stub needed for persistent"
+
+instance ToHttpApiData WalletId where
+    toUrlPiece = toText
+
+instance FromHttpApiData WalletId where
+    parseUrlPiece = parseUrlPieceFromText
+
+instance ToJSON WalletId where
+    toJSON = String . toText
+
+instance FromJSON WalletId where
+    parseJSON = aesonFromText "WalletId"
```

Comment:
What are those JSON & HttpApiData for ... in the DB layer ?

Reply:
they are needed to derive `Database.Persist.Class.PersistEntity.Key Wallet` and `Database.Persist.Class.PersistEntity.Key WalletPrivateKey`

Reply:
In the spirit of getting a SQL schema PR ready quickly for discussion, I put the orphan instances here. We can fix it in a later PR. Possibly by moving them to Primitive.Types.

Reply:
I still don't understand where they enter into play? Do `persistent` or `persistent-sqlite` requires explicit `Aeson` instances ?